overlord/snapshotstate/backend/reader.go

```diff
@@ -0,0 +1,374 @@
+	if err != nil {
+		rsh.Broken = err.Error()
+		return rsh, err
+	}
+	if sz.size != metaHashSize {
+		rsh.Broken = fmt.Sprintf("declared hash size (%d) does not match actual (%d)", metaHashSize, sz.size)
+		return rsh, errors.New(rsh.Broken)
+	}
+	if expectedMetaHash := string(bytes.TrimSpace(metaHashBuf)); actualMetaHash != expectedMetaHash {
+		logger.Noticef("Declared metadata hash (%s) does not match actual (%s).", expectedMetaHash, actualMetaHash)
+		rsh.Broken = "metadata hash mismatch"
+		return rsh, errors.New(rsh.Broken)
+	}
+
+	return rsh, nil
+}
+
+func (r *Reader) checkOne(ctx context.Context, entry string, hasher hash.Hash) error {
+	body, reportedSize, err := member(r.File, entry)
+	if err != nil {
+		return err
+	}
+	defer body.Close()
+
+	expectedHash := r.SHA3_384[entry]
+	readSize, err := io.Copy(io.MultiWriter(osutil.ContextWriter(ctx), hasher), body)
+	if err != nil {
+		return err
+	}
+
+	if readSize != reportedSize {
+		return fmt.Errorf("snapshot entry %q size (%d) different from actual (%d)", entry, reportedSize, readSize)
+	}
+
+	if actualHash := fmt.Sprintf("%x", hasher.Sum(nil)); actualHash != expectedHash {
+		logger.Noticef("Snapshot entry %q expected hash (%s) does not match actual (%s).", entry, expectedHash, actualHash)
+		return fmt.Errorf("snapshot entry %q expected hash different from actual", entry)
+	}
+	return nil
+}
+
+// Check that the data contained in the snapshot matches its hashsums.
+func (r *Reader) Check(ctx context.Context, usernames []string) error {
+	sort.Strings(usernames)
+
+	hasher := crypto.SHA3_384.New()
+	for entry := range r.SHA3_384 {
+		if len(usernames) > 0 && isUserArchive(entry) {
+			username := entryUsername(entry)
+			if !strutil.SortedListContains(usernames, username) {
+				logger.Debugf("In checking snapshot %q, skipping entry %q by user request.", r.Name(), username)
+				continue
+			}
+		}
+
+		if err := r.checkOne(ctx, entry, hasher); err != nil {
+			return err
+		}
+		hasher.Reset()
+	}
+
+	return nil
+}
+
+// A Logf writes somewhere the user will see it
+//
+// (typically this is task.Logf when running via the overlord, or fmt.Printf
+// when running direct)
+type Logf func(format string, args ...interface{})
+
+// Restore the data from the snapshot.
+//
+// If successful this will replace the existing data (for the revision
+// in the snapshot) with that contained in the snapshot.
+func (r *Reader) Restore(ctx context.Context, usernames []string, logf Logf) error {
+	_, err := r.restore(ctx, false, usernames, logf)
+	return err
+}
+
+// RestoreLeavingTrash is like Restore, but it leaves the old data in
+// a trash directory.
+func (r *Reader) RestoreLeavingTrash(ctx context.Context, usernames []string, logf Logf) (*Trash, error) {
+	return r.restore(ctx, true, usernames, logf)
+}
+
+func (r *Reader) restore(ctx context.Context, leaveTrash bool, usernames []string, logf Logf) (b *Trash, e error) {
+	b = &Trash{}
```

Comment:
heh. Because I'd called them backups originally (because they're a backup in the sense that `README~` is a backup of `README`, but you told me to rename it (so I called it trash, which is what we call it in snapstate).

Reply:
Let's please rename to the resulting type (RestoreState?) not Trash.. :P